templates/profile/solr/solr.in.sh.erb

```diff
@@ -0,0 +1,39 @@
+SOLR_JAVA_HOME="<%= @java_home %>"
+SOLR_HEAP="<%= @heap %>"
+
+GC_LOG_OPTS="-verbose:gc -XX:+PrintHeapAtGC -XX:+PrintGCDetails \
```

Comment:
I missed that HT had this enabled by default. It's much too verbose to always be on, and in general GC logging is something we typically only need for troubleshooting (Fulcrum needs it now, for example!). @botimer Can/should we make this something that is enabled only when needed based on a hiera flag?

Reply:
TL;DR: Add a parameter to the profile for supplemental JAVA_OPTS and expose this in the systemd unit as an environment variable. Append it to the default in the solr.in.sh script (with bash, not erb). Keep an external recipe book of options we might add for a given node via hieradata.

---

Where possible, we should push these cases toward a unified interface. That's essentially what projects like Tomcat and others have already done in their startup scripts with JAVA_OPTS. They don't know what you'll need or how you'll start things, or even what is supported by the JVM, but they know there is an environment where you can set a value they can put in the right spot.

This also aligns with systemd, where the unit files can hold environment variables directly, or point to an EnvironmentFile. The Twelve-Factor [config section](https://12factor.net/config) discusses this topic as well.

Expanding magic values into the script is better than nothing, but opaque. You can't tell from the concrete script whether a value is intrinsic to the package or injected by the management of the package. It's effectively a custom-built package for that node, and the variability is no longer contained. Think about doing a digest of a directory tree on disk -- the root of the install of the package is now machine-specific. It's better to keep that fixed and have it point to a much smaller, external area declared as volatile; the same as /usr/local vs. /etc.

So, what's that mean here?

We should make sockets in the script and plugs in the systemd unit. Giving special treatment to a couple of key items like max heap is fine, but it's ultimately a big JAVA_OPTS string. Breaking everything out is a lot more work and more likely to add confusion than clarity -- we don't want our management layer to create penalties for expertise with the thing being managed.

We should use puppet to install a fixed package and rely on systemd to set machine-specific environment variables. We can manage those variables either in the unit itself (already a template) or with an EnvironmentFile (a new, companion template).

This approach is all a subclass of separating the code (fixed) from the data (variable), and maximizing the portion of the system that can be held constant. It also favors common mechanisms over specialized, pushing detail outward -- the specialized stuff is transformed into standard stuff as far outside as possible, so the layers are progressively hardened, looking outside-in. In this case, hiera -> env file -> systemd unit -> solr -> jvm.

Reply:
Thanks for the thoughts and both long and short form information! That all seems reasonable to me. I imagine for the scope of the current pass we can delete this and introduce it when needed or when we apply the more general option-management solution you listed.

Reply:
This is fixed.